xarray/core/groupby.py

```diff
@@ -554,6 +509,44 @@ def factorize(self, squeeze: bool) -> None:
         self.codes = self.group.copy(data=codes)
 
 
+class Grouper(ABC):
+    pass
+
+
+class UniqueGrouper(Grouper):
+    pass
+
+
+class BinGrouper(Grouper):
+    def __init__(self, bins, cut_kwargs: Mapping | None):
+        if duck_array_ops.isnull(bins).all():
+            raise ValueError("All bin edges are NaN.")
+
+        if cut_kwargs is None:
+            cut_kwargs = {}
+
+        self.bins = bins
+        self.cut_kwargs = cut_kwargs
+
+
+class TimeResampleGrouper(Grouper):
+    def __init__(
+        self,
+        freq: str,
+        closed: SideOptions | None,
+        label: SideOptions | None,
+        origin: str | DatetimeLike,
+        offset: pd.Timedelta | datetime.timedelta | str | None,
+        loffset: datetime.timedelta | str | None,
+    ):
+        self.freq = freq
+        self.closed = closed
+        self.label = label
+        self.origin = origin
+        self.offset = offset
+        self.loffset = loffset
```

Comment:
This looks like a dataclass if you want to play around with that.

https://docs.python.org/3/library/dataclasses.html